src/WebJobs.Extensions.DurableTask/Listener/FunctionExecutionHelper.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.WebJobs.Host.Executors;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask.Listener
+{
+    internal static class FunctionExecutionHelper
+    {
+        public static async Task<WrappedFunctionResult> ExecuteFunctionInOrchestrationMiddleware(
+            ITriggeredFunctionExecutor executor,
+            IApplicationLifetimeWrapper hostServiceLifetime,
+            TriggeredFunctionData triggerInput,
+            CancellationToken cancellationToken)
+        {
+#pragma warning disable CS0618 // InvokeHandler approved for use by this extension
+            if (triggerInput.InvokeHandler == null)
+            {
+                throw new ArgumentException(
+                    $"{nameof(ExecuteFunctionInOrchestrationMiddleware)} should only be used when ${nameof(triggerInput)} has a value for ${nameof(TriggeredFunctionData.InvokeHandler)}");
+            }
+
+            try
+            {
+                bool executedUserCode = false;
+                var triggeredFunctionData = new TriggeredFunctionData()
+                {
+                    TriggerValue = triggerInput.TriggerValue,
+                    ParentId = triggerInput.ParentId,
+                    InvokeHandler = async userCodeHandler =>
+                    {
+                        executedUserCode = true;
+                        await triggerInput.InvokeHandler(userCodeHandler);
+                    },
+#if !FUNCTIONS_V1
+                    TriggerDetails = triggerInput.TriggerDetails,
+#endif
+                };
+#pragma warning restore CS0618
+
+                FunctionResult result = await executor.TryExecuteAsync(triggerInput, cancellationToken);
+                if (!result.Succeeded)
+                {
+                    if (!executedUserCode &&
+                         IsHostStopping(hostServiceLifetime))
+                    {
+                        return new WrappedFunctionResult(
+                            WrappedFunctionResult.FunctionResultStatus.FunctionsRuntimeError,
+                            result.Exception);
+                    }
+                    else
+                    {
+                        return new WrappedFunctionResult(
+                            WrappedFunctionResult.FunctionResultStatus.UserCodeError,
+                            result.Exception);
+                    }
+                }
+                else
+                {
+                    return new WrappedFunctionResult(
+                        WrappedFunctionResult.FunctionResultStatus.Success,
+                        null);
```

Comment:
I have a few suggestions for how to simplify this code. I'll batch them all into this one comment:

* Consider checking for the success case first instead of checking for the failure case first.
* Consider removing the `else` to reduce nesting.
* Consider creating static methods that represent each result type to further reduce code.
* Consider removing the `hostServiceLifetime` and using `cancellationToken` instead.
* Consider having a separate check for `!executeduserCode` that is not coupled to the host cancellation.

Here is an example that I think would be easier to read and understand:

```csharp
FunctionResult result = await executor.TryExecuteAsync(triggerInput, cancellationToken);
if (result.Succeeded)
{
    return WrappedFunctionResult.Success();
}

if (cancellationToken.IsCancellationRequested)
{
    return WrappedFunctionResult.FunctionsRuntimeError(result.Exception);
}

if (!executedUserCode)
{
   WrappedFunctionResult.UserCodeError(new InvalidOperationException(
        "The function failed to start executing. " +
        "For .NET functions, this can happen if an unhandled exception is thrown in the function's class constructor."));
}

return WrappedFunctionResult.UserCodeError(result.Exception);
```